src/runtime/uv/tcp.cpp

```diff
@@ -0,0 +1,332 @@
+
+        if (tcp_socket->m_promise_accept != NULL) {
+            lean_inc(f);
+            lean_apply_1(f, tcp_socket->m_promise_accept);
+        }
+
+        if (tcp_socket->m_promise_read != NULL) {
+            lean_inc(f);
+            lean_apply_1(f, tcp_socket->m_promise_read);
+        }
+
+        if (tcp_socket->m_byte_array != NULL) {
+            lean_inc(f);
+            lean_apply_1(f, tcp_socket->m_byte_array);
+        }
+    });
+}
+
+// TCP Socket Operations
+extern "C" LEAN_EXPORT lean_obj_res lean_uv_tcp_new() {
+    lean_uv_tcp_socket_object * tcp_socket = (lean_uv_tcp_socket_object*)malloc(sizeof(lean_uv_tcp_socket_object));
+
+    tcp_socket->m_promise_accept = NULL;
+    tcp_socket->m_promise_read = NULL;
+    tcp_socket->m_byte_array = NULL;
+
+    uv_tcp_t * uv_tcp = (uv_tcp_t*)malloc(sizeof(uv_tcp_t));
+
+    event_loop_lock(&global_ev);
+    int result = uv_tcp_init(global_ev.loop, uv_tcp);
+    event_loop_unlock(&global_ev);
+
+    if (result != 0) {
+        free(uv_tcp);
+        free(tcp_socket);
+        std::string err = std::string("failed to initialize tcp_socket: ") + uv_strerror(result);
+        return io_result_mk_error(err.c_str());
+    }
+
+    tcp_socket->m_uv_tcp = uv_tcp;
+
+    lean_object * obj = lean_uv_tcp_socket_new(tcp_socket);
+    lean_mark_mt(obj);
+
+    tcp_socket->m_uv_tcp->data = obj;
+
+    return lean_io_result_mk_ok(obj);
+}
+
+lean_object* create_promise() {
+    lean_object * prom_res = lean_io_promise_new(lean_io_mk_world());
+    lean_object * promise = lean_ctor_get(prom_res, 0);
+    lean_inc(promise);
+    lean_dec(prom_res);
+    return promise;
+}
+
+void resolve_promise(lean_object* promise, int status) {
+    lean_object * res;
+
+    if (status == 0) {
+        res = alloc_cnstr(1, 1, 0);
+        cnstr_set(res, 0, lean_box(0));
+    } else {
+        res = mk_except_error_string(uv_strerror(status));
+    }
+
+    lean_object * result = lean_io_promise_resolve(res, promise, lean_io_mk_world());
+    lean_dec(result);
+}
+
+extern "C" LEAN_EXPORT lean_obj_res lean_uv_tcp_connect(b_obj_arg socket, b_obj_arg addr) {
+    lean_uv_tcp_socket_object * tcp_socket = lean_to_uv_tcp_socket(socket);
+
+    sockaddr_in addr_ptr;
+    lean_socket_addr_to_sockaddr(addr, &addr_ptr);
+
+    lean_object * promise = create_promise();
+
+    uv_connect_t * uv_connect = (uv_connect_t*)malloc(sizeof(uv_connect_t));
+    uv_connect->data = promise;
+
+    event_loop_lock(&global_ev);
+
+    int result = uv_tcp_connect(uv_connect, tcp_socket->m_uv_tcp, (const struct sockaddr *)&addr_ptr, [](uv_connect_t* req, int status) {
+        lean_object * promise = (lean_object *) req->data;
+        resolve_promise(promise, status);
+        free(req);
+    });
+
+    if (result < 0) {
+        free(uv_connect);
+        return lean_io_result_mk_error(mk_except_error_string("Connection failed"));
+    }
+
+    event_loop_unlock(&global_ev);
+
+    return lean_io_result_mk_ok(promise);
```

Comment:
Afaict you are returning a promise with RC=1 to the user here and register a callback that can access it yes? Suppose the user drops the promise before the callback in `uv_tcp_connect` is called, you would do a use after free. It would appear to me you need to keep the refcount high here?